File: standalone/ios/VPinball/utils/VPinballManager.swift
import SwiftUI

class VPinballManager {
    static let shared = VPinballManager()

    var activeTable: Table?

    let impactGenerators: [UIImpactFeedbackGenerator.FeedbackStyle: UIImpactFeedbackGenerator] = [
        .heavy: UIImpactFeedbackGenerator(style: .heavy),
        .light: UIImpactFeedbackGenerator(style: .light),
        .medium: UIImpactFeedbackGenerator(style: .medium),
        .rigid: UIImpactFeedbackGenerator(style: .rigid),
        .soft: UIImpactFeedbackGenerator(style: .soft),
    ]

    let vpinballViewModel = VPinballViewModel.shared

    private init() {
        impactGenerators.forEach { $0.value.prepare() }
    }

    func startup() {
        VPinballInit { value, data in
            let vpinballManager = VPinballManager.shared
            let vpinballViewModel = VPinballViewModel.shared
            let event = VPinballEvent(rawValue: value)
            switch event {
            case .archiveCompressing,
                 .archiveUncompressing,
                 .loadingItems,
                 .loadingSounds,
                 .loadingImages,
                 .loadingFonts,
                 .loadingCollections,
                 .prerendering:
                if let data = data {
                    let json = String(cString: UnsafePointer<CChar>(data))
                    if let jsonData = json.data(using: .utf8),
                       let progressData = try? JSONDecoder().decode(ProgressEventData.self,
                                                                    from: jsonData)
                    {
                        let apply = {
                            if let name = event?.name {
                                vpinballViewModel.updateProgressHUD(progress: progressData.progress,
                                                                    status: name)
                            } else {
                                vpinballViewModel.updateProgressHUD(progress: progressData.progress)
                            }
                        }

                        if Thread.isMainThread {
                            apply()
                            CATransaction.flush()
                            RunLoop.main.run(mode: .default,
                                             before: Date().addingTimeInterval(0))
                        } else {
                            DispatchQueue.main.async {
                                apply()
                            }
                        }
                    }
                }
            case .playerStarted:
                vpinballViewModel.isPlaying = true
            case .rumble:
                if vpinballManager.loadValue(.standalone, "Haptics", true),
                   let data = data
                {
                    let json = String(cString: UnsafePointer<CChar>(data))
                    if let jsonData = json.data(using: .utf8),
                       let rumbleData = try? JSONDecoder().decode(RumbleData.self,
                                                                  from: jsonData)
                    {
                        vpinballManager.rumble(rumbleData)
                    }
                }
            case .scriptError:
                if vpinballViewModel.scriptError == nil,
                   let data = data
                {
                    let json = String(cString: UnsafePointer<CChar>(data))
                    if let jsonData = json.data(using: .utf8),
                       let scriptErrorData = try? JSONDecoder().decode(ScriptErrorData.self,
                                                                       from: jsonData)
                    {
                        let type = VPinballScriptErrorType(rawValue: CInt(scriptErrorData.error))!
                        vpinballViewModel.scriptError = "\(type.name) on line \(scriptErrorData.line), position \(scriptErrorData.position):\n\n\(scriptErrorData.description)"
                    }
                }
            case .playerClosed:
                vpinballViewModel.isPlaying = false
                vpinballManager.activeTable = nil

                if vpinballViewModel.scriptError != nil {
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                        vpinballViewModel.setAction(action: .showError,
                                                    table: nil)
                    }
                } else {
                    vpinballViewModel.setAction(action: .stopped)
                }
            case .webServer:
                if let data = data {
                    let json = String(cString: UnsafePointer<CChar>(data))
                    if let jsonData = json.data(using: .utf8),
                       let webServerData = try? JSONDecoder().decode(WebServerData.self,
                                                                     from: jsonData)
                    {
                        DispatchQueue.main.async {
                            vpinballViewModel.webServerURL = webServerData.url
                        }
                    }
                } else {
                    DispatchQueue.main.async {
                        vpinballViewModel.webServerURL = nil
                    }
                }
            case .tableListUpdated:
                var focusUuid: String?
                if let data = data {
                    let json = String(cString: UnsafePointer<CChar>(data))
                    if let jsonData = json.data(using: .utf8) {
                        struct TableListUpdatedData: Codable {
                            let focusUuid: String?
                        }
                        if let data = try? JSONDecoder().decode(TableListUpdatedData.self,
                                                                from: jsonData),
                            let uuidString = data.focusUuid
                        {
                            focusUuid = uuidString
                        }
                    }
                }

                DispatchQueue.main.async {
                    vpinballManager.activeTable = nil
                    vpinballViewModel.isPlaying = false
                    vpinballViewModel.hideHUD()

                    Task {
                        await TableManager.shared.loadTables()

                        if let uuid = focusUuid {
                            try? await Task.sleep(nanoseconds: 100_000_000)
                            vpinballViewModel.scrollToTableId = uuid
                        }
                    }
                }
            default:
                break
            }
        }
    }

    static func log(_ level: VPinballLogLevel, _ message: String) {
        VPinballLog(level.rawValue, message.cstring)
    }

    func loadValue(_ section: VPinballSettingsSection, _ key: String, _ defaultValue: CInt) -> CInt {
        return VPinballLoadValueInt(section.rawValue.cstring, key.cstring, defaultValue)
    }

    func loadValue(_ section: VPinballSettingsSection, _ key: String, _ defaultValue: Int) -> Int {
        return Int(loadValue(section, key, CInt(defaultValue)))
    }

    func loadValue(_ section: VPinballSettingsSection, _ key: String, _ defaultValue: Float) -> Float {
        return VPinballLoadValueFloat(section.rawValue.cstring, key.cstring, defaultValue)
    }

    func loadValue(_ section: VPinballSettingsSection, _ key: String, _ defaultValue: Bool) -> Bool {
        return loadValue(section, key, defaultValue ? 1 : 0) == 1
    }

    func loadValue(_ section: VPinballSettingsSection, _ key: String, _ defaultValue: String) -> String {
        return String(cString: VPinballLoadValueString(section.rawValue.cstring, key.cstring, defaultValue.cstring))
    }

    func saveValue(_ section: VPinballSettingsSection, _ key: String, _ value: CInt) {
        VPinballSaveValueInt(section.rawValue.cstring, key.cstring, value)
    }

    func saveValue(_ section: VPinballSettingsSection, _ key: String, _ value: Int) {
        saveValue(section, key, CInt(value))
    }

    func saveValue(_ section: VPinballSettingsSection, _ key: String, _ value: Float) {
        VPinballSaveValueFloat(section.rawValue.cstring, key.cstring, value)
    }

    func saveValue(_ section: VPinballSettingsSection, _ key: String, _ value: Bool) {
        saveValue(section, key, value ? 1 : 0)
    }

    func saveValue(_ section: VPinballSettingsSection, _ key: String, _ value: String) {
        VPinballSaveValueString(section.rawValue.cstring, key.cstring, value.cstring)
    }

    func rumble(_ data: RumbleData) {
        if data.lowFrequencyRumble > 0 || data.highFrequencyRumble > 0 {
            let style: UIImpactFeedbackGenerator.FeedbackStyle

            if data.lowFrequencyRumble == data.highFrequencyRumble {
                style = .rigid
            } else if data.lowFrequencyRumble > 20000 || data.highFrequencyRumble > 20000 {
                style = .heavy
            } else if data.lowFrequencyRumble > 10000 || data.highFrequencyRumble > 10000 {
                style = .medium
            } else if data.lowFrequencyRumble > 1000 || data.highFrequencyRumble > 1000 {
                style = .light
            } else {
                style = .soft
            }

            impactGenerators[style]!.impactOccurred()
        }
    }

    func play(table: Table) async -> Bool {
        if activeTable != nil {
            return false
        }

        if loadValue(.standalone, "ResetLogOnPlay", true) {
            VPinballResetLog()
        }

        activeTable = table
        vpinballViewModel.scriptError = nil

        await MainActor.run {
            vpinballViewModel.showProgressHUD(title: table.name,
                                              status: "Launching")
        }

        var success = true

        if await Task.detached(
            priority: .userInitiated,
            operation: { [table] in
                VPinballStatus(rawValue: VPinballLoadTable(table.uuid.cstring))
            }
        ).value == .success {
            if await MainActor.run(body: { VPinballStatus(rawValue: VPinballPlay()) }) != .success {
                try? await Task.sleep(nanoseconds: 500_000_000)

                activeTable = nil

                success = false

                VPinballManager.log(.error, "unable to start table")
            }
        } else {
            try? await Task.sleep(nanoseconds: 500_000_000)

            activeTable = nil

            success = false

            VPinballManager.log(.error, "unable to load table")
        }

        await MainActor.run {
            vpinballViewModel.hideHUD()
        }

        return success
    }

    func stop() {
        VPinballStop()
    }

    func resetIni() {
        _ = VPinballResetIni()
    }

    func updateWebServer() {
        VPinballUpdateWebServer()
    }
}

File: standalone/android/app/src/main/java/org/vpinball/app/VPinballManager.kt
package org.vpinball.app

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.util.Size
import java.io.File
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.koin.core.component.KoinComponent
import org.vpinball.app.jni.*
import org.vpinball.app.jni.VPinballSettingsSection.STANDALONE
import org.vpinball.app.ui.screens.landing.LandingScreenViewModel
import org.vpinball.app.util.FileUtils

object VPinballManager : KoinComponent {
    val vpinballJNI: VPinballJNI = VPinballJNI()

    internal lateinit var activity: VPinballActivity
    private lateinit var filesDir: File
    private lateinit var cacheDir: File
    private lateinit var displaySize: Size
    private lateinit var vibrator: Vibrator
    lateinit var safFileSystem: SAFFileSystem

    private var activeTable: Table? = null
    private var error: String? = null

    fun initialize(activity: VPinballActivity) {
        this.activity = activity
        this.safFileSystem = SAFFileSystem()

        filesDir = activity.filesDir
        cacheDir = activity.cacheDir

        val displayMetrics = activity.resources.displayMetrics
        val width = displayMetrics.widthPixels
        val height = displayMetrics.heightPixels
        displaySize = if (width > height) Size(height, width) else Size(width, height)

        vibrator =
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                val vibratorManager = activity.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
                vibratorManager.defaultVibrator
            } else {
                activity.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            }
    }

    fun startup() {
        runCatching { FileUtils.copyAssets(activity.assets, "", activity.filesDir) }

        vpinballJNI.VPinballInit { value, jsonData ->
            val viewModel = activity.viewModel
            val event = VPinballEvent.entries.find { it.ordinal == value }
            when (event) {
                VPinballEvent.ARCHIVE_UNCOMPRESSING,
                VPinballEvent.ARCHIVE_COMPRESSING,
                VPinballEvent.LOADING_ITEMS,
                VPinballEvent.LOADING_SOUNDS,
                VPinballEvent.LOADING_IMAGES,
                VPinballEvent.LOADING_FONTS,
                VPinballEvent.LOADING_COLLECTIONS,
                VPinballEvent.PRERENDERING -> {
                    val progressData =
                        jsonData?.let { jsonStr ->
                            try {
                                Json.decodeFromString<VPinballProgressData>(jsonStr)
                            } catch (e: Exception) {
                                log(VPinballLogLevel.WARN, "Failed to parse progress data JSON: $jsonStr - ${e.message}")
                                null
                            }
                        }
                    log(VPinballLogLevel.INFO, "event=${event.name}, data=${progressData}")
                    progressData?.let {
                        CoroutineScope(Dispatchers.Main).launch {
                            viewModel.title(activeTable?.name ?: "")
                            viewModel.progress(progressData.progress)
                            viewModel.status(event.text)
                        }
                    }
                }
                VPinballEvent.PLAYER_STARTED -> {
                    log(VPinballLogLevel.INFO, "event=${event.name}")
                    CoroutineScope(Dispatchers.Main).launch {
                        viewModel.playing(true)
                        delay(500)
                        viewModel.loading(false)
                    }
                }
                VPinballEvent.RUMBLE -> {
                    if (loadValue(STANDALONE, "Haptics", true)) {
                        val rumbleData =
                            jsonData?.let { jsonStr ->
                                try {
                                    Json.decodeFromString<VPinballRumbleData>(jsonStr)
                                } catch (e: Exception) {
                                    log(VPinballLogLevel.WARN, "Failed to parse rumble data JSON: $jsonStr - ${e.message}")
                                    null
                                }
                            }
                        rumbleData?.let { rumble(it) }
                    }
                }
                VPinballEvent.SCRIPT_ERROR -> {
                    log(VPinballLogLevel.INFO, "event=${event.name}, error=$error, jsonData=$jsonData")
                    if (error == null) {
                        val scriptErrorData =
                            jsonData?.let { jsonStr ->
                                try {
                                    Json.decodeFromString<VPinballScriptErrorData>(jsonStr)
                                } catch (e: Exception) {
                                    log(VPinballLogLevel.WARN, "Failed to parse script error data JSON: $jsonStr - ${e.message}")
                                    null
                                }
                            }
                        error =
                            scriptErrorData?.let {
                                val errorType = VPinballScriptErrorType.fromInt(it.error)
                                "${errorType.text} on line ${it.line}, position ${it.position}:\n\n${it.description}"
                            } ?: "Script error."
                        log(VPinballLogLevel.INFO, "SCRIPT_ERROR: set error to: $error")
                    } else {
                        log(VPinballLogLevel.INFO, "SCRIPT_ERROR: error already set, ignoring")
                    }
                }
                VPinballEvent.PLAYER_CLOSED -> {
                    log(VPinballLogLevel.INFO, "event=${event.name}")
                    activeTable = null
                    CoroutineScope(Dispatchers.Main).launch {
                        viewModel.playing(false)
                        viewModel.stopped()
                        error?.let { error ->
                            log(VPinballLogLevel.INFO, "STOPPED: showing error: $error")
                            delay(500)
                            showError(error)
                        } ?: log(VPinballLogLevel.INFO, "STOPPED: no error to show")
                    }
                }
                VPinballEvent.WEB_SERVER -> {
                    val webServerData =
                        jsonData?.let { jsonStr ->
                            try {
                                Json.decodeFromString<VPinballWebServerData>(jsonStr)
                            } catch (e: Exception) {
                                null
                            }
                        }
                    log(VPinballLogLevel.INFO, "event=${event.name}, data=${webServerData}")
                    webServerData?.let { CoroutineScope(Dispatchers.Main).launch { viewModel.webServerURL = webServerData.url } }
                }
                VPinballEvent.TABLE_LIST_UPDATED -> {
                    val focusUuid =
                        jsonData?.let { jsonStr ->
                            try {
                                @Serializable data class TableListUpdatedData(val focusUuid: String? = null)
                                val data = Json.decodeFromString<TableListUpdatedData>(jsonStr)
                                data.focusUuid
                            } catch (e: Exception) {
                                null
                            }
                        }

                    CoroutineScope(Dispatchers.Main).launch {
                        viewModel.loading(false)
                        LandingScreenViewModel.triggerRefresh()

                        focusUuid?.let { uuid ->
                            delay(100)
                            LandingScreenViewModel.triggerScrollToTable(uuid)
                        }
                    }
                }
                else -> {
                    log(VPinballLogLevel.WARN, "event=${event}")
                }
            }
        }
    }

    fun getDisplaySize(): Size {
        return displaySize
    }

    fun getFilesDir(): File {
        return filesDir
    }

    fun getCacheDir(): File {
        return cacheDir
    }

    fun log(level: VPinballLogLevel, message: String) {
        vpinballJNI.VPinballLog(level.value, message)
    }

    private fun rumble(data: VPinballRumbleData) {
        if (data.lowFrequencyRumble > 0 || data.highFrequencyRumble > 0) {
            val amplitude =
                when {
                    data.lowFrequencyRumble == data.highFrequencyRumble -> VibrationEffect.DEFAULT_AMPLITUDE
                    data.lowFrequencyRumble > 20000 || data.highFrequencyRumble > 20000 -> 255
                    data.lowFrequencyRumble > 10000 || data.highFrequencyRumble > 10000 -> 200
                    data.lowFrequencyRumble > 1000 || data.highFrequencyRumble > 1000 -> 100
                    else -> 50
                }
            vibrator.vibrate(VibrationEffect.createOneShot(15, amplitude))
        }
    }

    fun updateWebServer() {
        vpinballJNI.VPinballUpdateWebServer()
    }

    fun getVersionString(): String = vpinballJNI.VPinballGetVersionStringFull()

    fun loadValue(section: VPinballSettingsSection, key: String, defaultValue: Int): Int =
        vpinballJNI.VPinballLoadValueInt(section.value, key, defaultValue)

    fun loadValue(section: VPinballSettingsSection, key: String, defaultValue: Float): Float =
        vpinballJNI.VPinballLoadValueFloat(section.value, key, defaultValue)

    fun loadValue(section: VPinballSettingsSection, key: String, defaultValue: Boolean): Boolean =
        loadValue(section, key, if (defaultValue) 1 else 0) == 1

    fun loadValue(section: VPinballSettingsSection, key: String, defaultValue: String): String =
        vpinballJNI.VPinballLoadValueString(section.value, key, defaultValue)

    fun saveValue(section: VPinballSettingsSection, key: String, value: Int) {
        vpinballJNI.VPinballSaveValueInt(section.value, key, value)
    }

    fun saveValue(section: VPinballSettingsSection, key: String, value: Float) {
        vpinballJNI.VPinballSaveValueFloat(section.value, key, value)
    }

    fun saveValue(section: VPinballSettingsSection, key: String, value: Boolean) {
        saveValue(section, key, if (value) 1 else 0)
    }

    fun saveValue(section: VPinballSettingsSection, key: String, value: String) {
        vpinballJNI.VPinballSaveValueString(section.value, key, value)
    }

    fun resetIni() {
        vpinballJNI.VPinballResetIni()
    }

    fun getVPinballCacheDir(): File = cacheDir

    fun play(table: Table) {
        if (activeTable != null) return
        activeTable = table
        error = null
        CoroutineScope(Dispatchers.IO).launch {
            if (loadValue(STANDALONE, "ResetLogOnPlay", true)) {
                vpinballJNI.VPinballResetLog()
            }
            withContext(Dispatchers.Main) { activity.viewModel.loading(true, table) }
            if (vpinballJNI.VPinballLoadTable(table.uuid) == VPinballStatus.SUCCESS.value) {
                vpinballJNI.VPinballPlay()
            } else {
                delay(500)
                withContext(Dispatchers.Main) {
                    activity.viewModel.stopped()
                    showError("Unable to load table.")
                    activeTable = null
                }
            }
        }
    }

    fun stop() {
        vpinballJNI.VPinballStop()
    }

    fun fileExists(path: String): Boolean {
        return vpinballJNI.VPinballFileExists(path)
    }

    fun stageFile(path: String): String? {
        return vpinballJNI.VPinballStageFile(path)
    }

    fun showError(message: String) {
        CoroutineScope(Dispatchers.Main).launch {
            delay(250)
            activity.viewModel.setError(message)
        }
    }

    fun getTablesPath(): String {
        return vpinballJNI.VPinballGetTablesPath()
    }

    fun getCurrentTablesPath(): String {
        return vpinballJNI.VPinballGetTablesPath()
    }

    fun isTablesPathSAF(): Boolean {
        return getTablesPath().startsWith("content://")
    }

    fun logAvailableStoragePaths() {
        log(VPinballLogLevel.INFO, "=== TESTING getExternalFilesDirs() ===")
        val externalDirs = activity.getExternalFilesDirs(null)
        log(VPinballLogLevel.INFO, "Found ${externalDirs.size} external storage locations:")
        externalDirs.forEachIndexed { index, file ->
            if (file != null) {
                log(VPinballLogLevel.INFO, "  [$index] ${file.absolutePath}")
                log(VPinballLogLevel.INFO, "       Exists: ${file.exists()}, CanWrite: ${file.canWrite()}")
            } else {
                log(VPinballLogLevel.INFO, "  [$index] null")
            }
        }
        log(VPinballLogLevel.INFO, "=== END STORAGE TEST ===")
    }

    fun RefreshTables() {
        log(VPinballLogLevel.INFO, "VPinballManager: RefreshTables() called")
        CoroutineScope(Dispatchers.IO).launch {
            val status = vpinballJNI.VPinballRefreshTables()
            if (status == VPinballStatus.SUCCESS.value) {
                log(VPinballLogLevel.INFO, "VPinballManager: Successfully reloaded tables path")
            } else {
                log(VPinballLogLevel.ERROR, "VPinballManager: Failed to reload tables path - status: $status")
            }
        }
    }
}
